src/crab_verifier.cpp

```diff
@@ -99,43 +101,33 @@ static checks_db generate_report(cfg_t& cfg,
     return m_db;
 }
 
-template<typename F>
-auto timed_execution(F f) {
-    clock_t begin = clock();
-
-    const auto& res = f();
-
-    clock_t end = clock();
-
-    double elapsed_secs = double(end - begin) / CLOCKS_PER_SEC;
-    return std::make_tuple(res, elapsed_secs);
-}
-
-static void print_report(const checks_db& db) {
-    std::cout << "\n";
+static void print_report(std::ostream& s, const checks_db& db) {
+    s << "\n";
     for (auto [label, messages] : db.m_db) {
-        std::cout << label << ":\n";
+        s << label << ":\n";
         for (const auto& msg : messages)
-            std::cout << "  " << msg << "\n";
+            s << "  " << msg << "\n";
     }
-    std::cout << "\n";
-    std::cout << db.total_warnings << " warnings\n";
+    s << "\n";
+    s << db.total_warnings << " warnings\n";
 }
 
-/// First item of returned value is true if the program passes verification
-std::tuple<bool, double> run_ebpf_analysis(cfg_t& cfg, program_info info) {
+/// Returned value is true if the program passes verification.
+bool run_ebpf_analysis(std::ostream& s, cfg_t& cfg, program_info info, const ebpf_verifier_options_t* options) {
+    if (options == nullptr)
+        options = &ebpf_verifier_default_options;
+
     global_program_info = std::move(info);
     crab::domains::clear_global_state();
 
-    auto&& [report, elapsed_secs] = timed_execution([&] {
-        // Get dictionaries of preconditions and postconditions for each
-        // basic block.
-        auto [preconditions, postconditions] = crab::run_forward_analyzer(cfg);
-        return generate_report(cfg, preconditions, postconditions);
-    });
+    // Get dictionaries of preconditions and postconditions for each
+    // basic block.
+    auto [preconditions, postconditions] = crab::run_forward_analyzer(cfg);
+
+    checks_db report = generate_report(s, cfg, preconditions, postconditions, *options);
 
-    if (global_options.print_failures) {
-        print_report(report);
+    if (options->print_failures) {
+        print_report(s, report);
     }
```

Comment:
Printing is the main reason I put the measurements inside this function: it's meaningless to have both verbose mode and measurements. I'm not sure that's important though, since measurements are realtively consistent across runs.

Reply:
That's what I concluded too.  Is there a change you'd like me to make or is it ok?

Reply:
No, I think that's ok.